### PR TITLE
xbar: init at 2.1.7-beta

### PR DESCRIPTION
**Pull request by r17x**
https://xbarapp.com/

###### Tested Compilations
```console
1 package added:
xbar (init at 2.1.7-beta)

1 package built:
xbar
```

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r17x** [comment]
@Janik-Haag I read from [maintainer-list.nix](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix#L30) about invitation to the orgs. How to get invitation?

**Comment by Janik-Haag** [comment]
Hi, a bot invites you automatically once your maintainers entry object gets merged.
I have no time to review this pr, sorry (the label was added by [a bot](https://github.com/nix-community/first-time-contribution-tagger/)). You can ask on discourse or matrix for reviews.

File: maintainers/maintainer-list.nix
Comment (hraban) [inline maintainers/maintainer-list.nix:15766]: I think this is supposed to go in a separate PR? Or is that old information I have?

File: pkgs/by-name/xb/xbar/package.nix
Comment (hraban) [inline pkgs/by-name/xb/xbar/package.nix:8]: Afaik this is undesirable for nixpkgs, particularly for open source software. xbar is open source and released under the MIT license, I would assume we want to compile it from source here.

**Review by hraban — COMMENTED** [review]
Personally keen on seeing this happen but I don't think a binary distribution is the way for nixpkgs. My 2¢ fwiw.

**Comment by hraban** [comment]
Btw I had a look at how to compile this on OSX and apparently it requires nodejs (ok), go (ok), and "wails"? some kind of golang library which lets you write osx guis in go & html? It exists in nixpkgs, but currently linux only, which is *ironic* to say the least.

File: maintainers/maintainer-list.nix
Comment (Janik-Haag) [reply maintainers/maintainer-list.nix:15766]: Same PR is fine, but it should be a separate commit called "maintainers: add r17x` since we try to keep a atomic commit history in nixpkgs

File: pkgs/by-name/xb/xbar/package.nix
Comment (Janik-Haag) [reply pkgs/by-name/xb/xbar/package.nix:8]: Yes compiling from source is always preferable

File: pkgs/by-name/xb/xbar/package.nix
Comment (Janik-Haag) [inline pkgs/by-name/xb/xbar/package.nix:6]:
```suggestion
  version = "2.1.7-beta";
  
  src = fetchFromGitHub {
```
just some formatting, and if you go the way of building from source (which you should) then you can switch to the fetchFromGitHub wrapper

**Review by hraban — COMMENTED** [review]
I'm going to change my original feedback about compiling from source being preferable on the basis that one of the dependencies for this isn't available for darwin in nixpkgs, it only works on linux. I don't think it's fair to make this PR contingent on fixing that, as well. It would be nice to add a comment, maybe?

"Downloading as binary package for now because compiling from source requires wails, which is currently only available for linux in nixpkgs."

What do y'all think?

**Review by hraban — APPROVED** [review]
I know it's old but it still builds and there are no merge conflicts. I just merged this into master locally, built and tested it, and it worked fine.

**Review by thomedw — APPROVED** [review]

File: maintainers/maintainer-list.nix
Comment (r17x) [reply maintainers/maintainer-list.nix:16854]: Done @Janik-Haag

File: pkgs/by-name/xb/xbar/package.nix
Comment (r17x) [reply pkgs/by-name/xb/xbar/package.nix]: Done

File: pkgs/by-name/xb/xbar/package.nix
Comment (r17x) [reply pkgs/by-name/xb/xbar/package.nix:19]: What should I need to add in preInstall and postInstall

File: maintainers/maintainer-list.nix
Comment (Janik-Haag) [reply maintainers/maintainer-list.nix:16852]: I quit nixpkgs and won't be reviewing this any further #323901

File: pkgs/by-name/xb/xbar/package.nix
Comment (hraban) [reply pkgs/by-name/xb/xbar/package.nix:25]: @Janik-Haag means:
```suggestion
  installPhase = ''
    runHook preInstall

    mkdir -p $out/Applications/xbar.app
    cp -R . $out/Applications/xbar.app
    
    runHook postInstall
  '';
```

File: maintainers/maintainer-list.nix
Comment (r17x) [reply maintainers/maintainer-list.nix:16852]: Thank for your contributions  @Janik-Haag

File: pkgs/by-name/xb/xbar/package.nix
Comment (r17x) [reply pkgs/by-name/xb/xbar/package.nix:25]: Updated!

File: pkgs/by-name/xb/xbar/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/xb/xbar/package.nix:6]:
```suggestion
```

File: pkgs/by-name/xb/xbar/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/xb/xbar/package.nix:37]:
```suggestion
  meta = {
    description = "Put the output from any script or program into your macOS Menu Bar (the BitBar reboot)";
    homepage = "https://xbarapp.com/";
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
    platforms = lib.platforms.darwin;
    maintainers = with lib.maintainers; [ r17x ];
    license = lib.licenses.mit;
```

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

File: pkgs/by-name/xb/xbar/package.nix
Comment (r17x) [reply pkgs/by-name/xb/xbar/package.nix]: Done

File: pkgs/by-name/xb/xbar/package.nix
Comment (r17x) [reply pkgs/by-name/xb/xbar/package.nix]: Done!

**Comment by hraban** [comment]
This PR still works, would love to see this get merged in. Thanks for your patience @r17x if I had merge rights I'd do it but alas.

**Comment by NetaliDev** [comment]
I just checked and apparently `wails` is now available for Darwin after all. So it should be possible to build it from source by now.

**Comment by hraban** [comment]
> I just checked and apparently `wails` is now available for Darwin after all. So it should be possible to build it from source by now.

I don't know whether to laugh or cry... 😂😭 poor @r17x . This PR deserves a place in a Hall of some kind. 

I guess the rule is to call this xbar-bin, then, right? Would that be an acceptable PR? I can't bear to see this rejected after 2 years because we waited so long to merge it that the build ecosystem has changed from under it 😂

**Comment by r17x** [comment]
Lemme make it build from Source @NetaliDev , Maybe I needed some help when I stuck.

Don't worries @hraban,  I will cry & laugh at the same time.

**Comment by NetaliDev** [comment]
Sorry for my wording at this point. I know how frustrating it must be, but unfortunately it is a fact that source builds are preferred in the nixpkgs. And of course it's very unfortunate that the PR has been open for so long that a source build is now possible again, but as frustrating as it is, I think we should at least try it. And of course, if it turns out to be too complicated for whatever reason, we can stick with the binaries.

**Comment by r17x** [comment]
@NetaliDev Don't worries. Let's make it build from source. 

```nix
{
  lib,
  fetchFromGitHub,
  buildGoModule,
  wails,
  pkg-config,
  autoPatchelfHook,
}:

let
  version = "2.1.7-beta";
  src = fetchFromGitHub {
    owner = "matryer";
    repo = "xbar";
    rev = "v${version}";
    sha256 = "sha256-RiTJgMqj5CRykvLCsqdioP8iB+aO8BQMm71KQ3WB75k=";
  };

in

buildGoModule {
  name = "xbar";
  pname = "xbar";

  inherit src;
  vendorHash = null;
  modRoot = "app";
  goDeps = src + "/app/go.sum";

  nativeBuildInputs = [
    wails
    pkg-config
    autoPatchelfHook
  ];

  patchPhase = ''
    echo -n "v${version}" > .version
  '';

  buildPhase = ''
    runHook preBuild
    wails build -m -s -trimpath -skipbindings  -o xbar
    runHook postBuild
  '';

  installPhase = ''
    runHook preInstall

    mkdir -p $out/Applications
    cp -R $out/xbar $out/Applications/xbar.app

    runHook postInstall
  '';

  meta = {
    description = "Put the output from any script or program into your macOS Menu Bar (the BitBar reboot)";
    homepage = "https://xbarapp.com/";
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
    platforms = lib.platforms.darwin;
    maintainers = with lib.maintainers; [ r17x ];
    license = lib.licenses.mit;
  };
}
```

Anyway, I have try to build from source, got an error with `buildGoModule`. I think, I missed something

```log
       >                To ignore the vendor directory, use -mod=readonly or -mod=mod.
       >                 To sync the vendor directory, run:
       >                     go mod vendor

```

**Comment by NetaliDev** [comment]
Okay, I have also tried to build it from source and I gave up. I can't even build it outside a derivation by running the provided build scripts. And even if we could build it: I bet that macOS will complain because it is not signed.

I've never had anything to do with macOS packaging under Nix, but when I grep for `undmg` in the nixpkgs, I find many macOS apps that could actually be built from source, but are simply unpacked as binary packages and I would also suspect that the signatures are a reason for this, but I'm not sure.

Anyway, if all the other apps are allowed to exist as binary packages, this one should be able to as well, so let's just go with the binary package.

**Review by NetaliDev — APPROVED** [review]

**Pull request merged**
